consumers/campaigner/main.go

```diff
@@ -38,165 +42,203 @@ func (h *MessageHandler) HandleMessage(m *nsq.Message) error {
 	}
 
 	msg := new(entities.CampaignerTopicParams)
+	svc := campaigns.New(h.s, h.p)
 
 	err := json.Unmarshal(m.Body, msg)
 	if err != nil {
 		logrus.WithField("body", string(m.Body)).WithError(err).Error("Malformed JSON message.")
 		return nil
 	}
 
-	c, err := h.s.GetCampaign(msg.CampaignID, msg.UserID)
+	campaign, err := h.s.GetCampaign(msg.CampaignID, msg.UserID)
 	if err != nil {
+		if errors.Is(err, gorm.ErrRecordNotFound) {
+			return nil
+		}
 		logrus.WithFields(logrus.Fields{
 			"campaign_id": msg.CampaignID,
-			"user_id":     msg.UserID,
-		}).WithError(err).Error("Unable to fetch campaign")
+			"user_id":     msg.UserID},
+		).WithError(err).Error("unable to find campaign")
+		return err
+	}
+	if campaign.Status != entities.StatusDraft {
+		logrus.WithError(err).
+			WithFields(logrus.Fields{
+				"user_id":     msg.UserID,
+				"campaign_id": msg.CampaignID,
+			}).
+			Errorf("potentially duplicate message: campaign status is '%s', it should be 'draft'", campaign.Status)
+		return nil
+	}
+
+	campaign.StartedAt.SetValid(time.Now().UTC())
+	campaign.Status = entities.StatusSending
+	err = h.s.UpdateCampaign(campaign)
+	if err != nil {
+		logrus.WithError(err).
+			WithFields(logrus.Fields{
+				"user_id":     campaign.UserID,
+				"campaign_id": campaign.ID,
+				"status":      campaign.Status,
+			}).
+			Error("unable to update campaign")
 		return nil
 	}
 
-	// fetching subs that are active and that have not been blacklisted
 	var (
-		active      = true
-		blacklisted = false
-		timestamp   time.Time
-		nextID      int64
-		limit       int64 = 1000
+		timestamp time.Time
+		nextID    int64
+		limit     int64 = 1000
 	)
+
+	template, err := h.svc.GetTemplate(context.Background(), campaign.TemplateID, msg.UserID)
+	if err != nil {
+		if errors.Is(err, gorm.ErrRecordNotFound) {
+			logrus.WithError(err).
+				WithFields(logrus.Fields{
+					"template_id": campaign.TemplateID,
+					"user_id":     msg.UserID,
+				}).
+				Error("unable to get template")
+			return nil
+		}
+		logrus.WithError(err).
+			WithFields(logrus.Fields{
+				"template_id": campaign.TemplateID,
+				"user_id":     msg.UserID,
+			}).
+			Error("unable to get template")
+		return err
+	}
+
+	html, err := mustache.ParseString(template.HTMLPart)
```

Comment:
@FilipNikolovski  @Dzalevski Maybe we should handle these errors differently change the status to the campaign as failed or somehow tell the user that its a problem with the template idk.

Reply:
Yes, but technically this would rarely happen since we already validate the templates in the endpoint. I'm not sure how we can notify the user, except store campaign failed events and show them to the user later on.

Reply:
@gudgl @FilipNikolovski what if update fails? :D

Reply:
we should store failed campaign on all errors right? @FilipNikolovski @gudgl